File: dashr/chapters/callbacks/examples/hello-slider.R
library(dashR)
library(dashCoreComponents)
library(dashHtmlComponents)

app <- Dash$new()

df <- read.csv(url("https://raw.githubusercontent.com/plotly/datasets/master/gapminderDataFiveYear.csv"))

continents <- unique(df$continent)
years <- unique(df$year)

# dccSlider starts from 0;
app$layout(
  htmlDiv(
    list(
      dccGraph(id = 'graph-with-slider'),
      dccSlider(
        id = 'year-slider',
        min = 0,
        max = length(years) - 1,
        marks = years,
        value = 0
      )
    )
  )
)



app$callback(
  output = list(id='graph-with-slider', property='figure'),
  params = list(input(id='year-slider', property='value')),
  
  function(indice_selected_year) {
    
    which_year_is_selected <- which(df$year == years[indice_selected_year + 1])
    
    traces <- lapply(continents,
                     function(cont) {
                       
                       which_continent_is_selected <- which(df$continent == cont)
                       
                       df_sub <- df[intersect(which_year_is_selected, which_continent_is_selected), ]
                       
                       with(
                         df_sub, 
                         list(
                           x = gdpPercap,
                           y = lifeExp,
                           opacity=0.5,
                           text = country,
                           mode = 'markers',
                           marker = list(
                             size = 15,
                             line = list(width = 0.5, color = 'white')
                           ),
                           name = cont
                         )
                       )
                     }
    )
    
    list(
      data = traces,
      layout= list(
        xaxis = list(type = 'log', title = 'GDP Per Capita'),
        yaxis = list(title = 'Life Expectancy', range = c(20,90)),
        margin = list(l = 40, b = 40, t = 10, r = 10),
        legend = list(x = 0, y = 1),
        hovermode = 'closest'
      )
    )
  }
)

app$run_server()

File: dash_docs/chapters/basic_callbacks/examples/hello-slider.jl
using CSV, DataFrames, Dash, DashHtmlComponents, DashCoreComponents


url = "https://raw.githubusercontent.com/plotly/datasets/master/gapminderDataFiveYear.csv"
download(url, "gapminder-data.csv")
df = DataFrame(CSV.File("gapminder-data.csv"))

continents = unique(df[!, :continent])
years = unique(df[!, :year])

app = dash()

app.layout = html_div() do
    dcc_graph(id="graph"),
    dcc_slider(
        id = "year-slider",
        min = 0,
        max = length(years) - 1,
        marks = years,
        value = 0)
end

callback!(
    app,
    Output("graph", "figure"),
    Input("year-slider", "value"),
) do index
    single_year_df = df[df.year .== years[index+1], :]

    figure_data = []

    for cont in continents
        single_continent_df = single_year_df[single_year_df.continent .== cont, :]
        push!(figure_data,
             (x = single_continent_df[:, :gdpPercap],
              y = single_continent_df[:, :lifeExp],
              type = "scatter",
              mode = "markers",
              hovertext = single_continent_df[:, :country],
              opacity=0.5,
              name = cont)
        )
    end

    figure = (
        data = figure_data,
        layout = (xaxis = ((type="log"), (title="GDP")),
                  yaxis = ((title="Life Expectancy"), range=(20, 90)),
                  transition_duration=500
        ),
    )

    return figure
end

run_server(app, "0.0.0.0", 8000)
